File: doodads/Tests/VBNet/Oracle/OracleAggregateFixture.vb
Imports System
Imports System.Data
Imports NUnit.Framework
Imports MyGeneration.dOOdads
Imports Npgsql

Namespace MyGeneration.dOOdads.Tests.Oracle

	<TestFixture()> _
	 Public Class FirebirdAggregateFixture

		Dim aggTest As AggregateTest = New AggregateTest

		<TestFixtureSetUp()> _
		Public Sub Init()
			TransactionMgr.ThreadTransactionMgrReset()
			aggTest.ConnectionString = Connections.OracleConnection
			UnitTestBase.RefreshDatabase()
		End Sub

		<SetUp()> _
		Public Sub Init2()
			aggTest.FlushData()
		End Sub

		<Test()> _
		Public Sub EmptyQueryReturnsSELLECTAll()
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(30, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateAvg()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Avg
			aggTest.Aggregate.Salary.Alias = "Avg"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateCount()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Count
			aggTest.Aggregate.Salary.Alias = "Count"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateMin()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Min
			aggTest.Aggregate.Salary.Alias = "Min"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateMax()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Max
			aggTest.Aggregate.Salary.Alias = "Max"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateSum()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Sum
			aggTest.Aggregate.Salary.Alias = "Sum"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateStdDev()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.StdDev
			aggTest.Aggregate.Salary.Alias = "Std Dev"
			aggTest.Query.Load()
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddAggregateVar()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Var
			aggTest.Aggregate.Salary.Alias = "Var"
			aggTest.Query.Load()
		End Sub

		<Test()> _
		Public Sub AddAggregateCountAll()
			aggTest.Query.CountAll = True
			aggTest.Query.CountAllAlias = "Total"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AddTwoAggregates()
			aggTest.Query.CountAll = True
			aggTest.Query.CountAllAlias = "Total"
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Sum
			aggTest.Aggregate.Salary.Alias = "Sum"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AggregateWithTearoff()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Sum
			aggTest.Aggregate.Salary.Alias = "Sum"
			Dim ap As AggregateParameter = aggTest.Aggregate.TearOff.Salary
			ap.Function = AggregateParameter.Func.Min
			ap.Alias = "Min"
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub AggregateWithWhere()
			aggTest.Query.CountAll = True
			aggTest.Query.CountAllAlias = "Total"
			aggTest.Where.IsActive.Operator = WhereParameter.Operand.Equal
			aggTest.Where.IsActive.Value = True
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual(1, aggTest.RowCount)
		End Sub

		<Test()> _
		Public Sub EmptyAliasUsesColumnName()
			aggTest.Aggregate.Salary.Function = AggregateParameter.Func.Sum
			Assert.IsTrue(aggTest.Query.Load())
			Assert.AreEqual("SALARY", aggTest.Aggregate.SALARY.Alias)
		End Sub

		<Test()> _
		Public Sub DistinctAggregate()
			aggTest.Aggregate.LastName.Function = AggregateParameter.Func.Count
			aggTest.Aggregate.LastName.Distinct = True
			Assert.IsTrue(aggTest.Query.Load())
			Dim testTable As DataTable = aggTest.DefaultView.Table
			Dim currRows As DataRow() = testTable.Select("", "", DataViewRowState.CurrentRows)
			Dim testRow As DataRow = currRows(0)
			Assert.AreEqual(10, testRow(0))
		End Sub

	End Class

End Namespace

File: doodads/Tests/CSharp/Oracle/OracleAggregateFixture.cs
using System;
using System.Data;
using NUnit.Framework;
using MyGeneration.dOOdads;

namespace MyGeneration.dOOdads.Tests.Oracle
{
	[TestFixture]
	public class OracleAggregateFixture
	{
		Oracle.AggregateTest aggTest = new Oracle.AggregateTest();
		
		[TestFixtureSetUp]
		public void Init()
		{
			aggTest.ConnectionStringConfig = "OracleConnection";
			UnitTestBase.RefreshDatabase();
		}

		[SetUp]
		public void Init2()
		{
			aggTest.FlushData();
		}

		[Test]
		public void EmptyQueryReturnsSELLECTAll()
		{
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(30, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateAvg()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Avg;
			aggTest.Aggregate.SALARY.Alias = "Avg";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateCount()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Count;
			aggTest.Aggregate.SALARY.Alias = "Count";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateMin()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Min;
			aggTest.Aggregate.SALARY.Alias = "Min";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateMax()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Max;
			aggTest.Aggregate.SALARY.Alias = "Max";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateSum()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Sum;
			aggTest.Aggregate.SALARY.Alias = "Sum";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateStdDev()
		{
			aggTest.Aggregate.AGE.Function = AggregateParameter.Func.StdDev;
			aggTest.Aggregate.AGE.Alias = "Std Dev";
			Assert.IsTrue(aggTest.Query.Load());
		    Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateVar()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Var;
			aggTest.Aggregate.SALARY.Alias = "Var";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddAggregateCountAll()
		{
			aggTest.Query.CountAll = true;
			aggTest.Query.CountAllAlias = "Total";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AddTwoAggregates()
		{
			aggTest.Query.CountAll = true;
			aggTest.Query.CountAllAlias = "Total";
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Sum;
			aggTest.Aggregate.SALARY.Alias = "Sum";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AggregateWithTearoff()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Sum;
			aggTest.Aggregate.SALARY.Alias = "Sum";
			AggregateParameter ap = aggTest.Aggregate.TearOff.SALARY;
			ap.Function = AggregateParameter.Func.Min;
			ap.Alias = "Min";
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}

		[Test]
		public void AggregateWithWhere()
		{
			aggTest.Query.CountAll = true;
			aggTest.Query.CountAllAlias = "Total";
			aggTest.Where.ISACTIVE.Operator = WhereParameter.Operand.Equal;
			aggTest.Where.ISACTIVE.Value = true;
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual(1, aggTest.RowCount);
		}
		
		[Test]
		public void EmptyAliasUsesColumnName()
		{
			aggTest.Aggregate.SALARY.Function = AggregateParameter.Func.Sum;
			Assert.IsTrue(aggTest.Query.Load());
			Assert.AreEqual("SALARY", aggTest.Aggregate.SALARY.Alias);
		}
		
		[Test]
		public void DistinctAggregate()
		{
			aggTest.Aggregate.LASTNAME.Function = AggregateParameter.Func.Count;
			aggTest.Aggregate.LASTNAME.Distinct = true;
			Assert.IsTrue(aggTest.Query.Load());
			DataTable testTable = aggTest.DefaultView.Table;
			DataRow[] currRows = testTable.Select(null, null, DataViewRowState.CurrentRows);
			DataRow testRow = currRows[0];
			Assert.AreEqual(10, testRow[0]);
		}

	}
}
